### PR TITLE
Create dynamic node get_* functions, based off of with_* functions.

### PR DESCRIPTION
**Pull request by alexjmacdonald**
### Which issue does this PR correspond to?

#233 

### What changes does this PR make to Grapl? Why?

When deriving as a Dynamic Node, grapl creates `with_[property_name](property_value)` functions, allowing a builder-style syntax.  This is an alternative to calling `set_property("[property_name]", property_value)` that pushes the compile-time index check to the compiler, providing a compile-time check that the calling author indeed is referencing an expectably extant value.

This PR extends the syntax for "getter" functions.  The example was extended to reflect intended usage.

### How were these changes tested?

`cargo check` and visual inspection.  Almost all of this code is macro or compile-time checks, so I assume that if the example code passes, this is good.

### CODE REVIEW

File: src/rust/graph-descriptions/src/lib.rs [hunk 1]
```diff
@@ -62,6 +62,27 @@ impl From<u64> for NodeProperty {
...
 }
 
+impl From<NodeProperty> for Option<String> {
+    fn from(np: NodeProperty) -> Option<String> {
+        match np.as_str_prop() {
+            Some(str) => Some(String::from(str)),
+            None => None,
+        }
+    }
+}
```
Comment (colin-grapl) [inline src/rust/graph-descriptions/src/lib.rs:72]: We can avoid an allocation (String::from allocates) like this:
```
impl From<NodeProperty> for Option<String> {
    fn from(np: NodeProperty) -> Option<String> {
        match np {
            NodeProperty {
                property: Some(node_property::Property::Strprop(s))
            } => Some(s),
           _ => None,
        }
    }
}
```

**Review by colin-grapl — COMMENTED** [review]
A bit of feedback on some implementation details, and overall I think I might suggest that during the codegen we determine the type, and then generate a different body for the `get_method` based on that type. I provided example code for the String version of that.

**Comment by wimax-grapl** [comment]
also fyi ignore the e2e failure :)

File: src/rust/derive-dynamic-node/src/lib.rs [hunk 2]
```diff
@@ -221,6 +219,14 @@ fn get_method(property_name: &Ident, property_type: &Type) -> TS2 {
             );
             self
         }
+
+        fn #get_method_name(&self) -> Option<#property_type> {
+            let property_result: Option<&NodeProperty> = self.get_dynamic_node().get_property(#property_name_str);
+            match property_result {
+                Some(property) => Option::< #property_type >::from(property.clone()),
+                None => None
```
[the remark this answers is not in the document]
Comment (alexjmacdonald) [reply src/rust/derive-dynamic-node/src/lib.rs:227]: Ok, I have a better idea of what's going on here.  This works:

```
fn #get_method_name(&self) -> Option<&str> {
    let property_result: Option<&NodeProperty> = self.get_dynamic_node().get_property(#property_name_str);

    match property_result {
      Some(ref property) => property.as_str_prop(),
      None => None,
    }
}
```

Now, `String` fields map to `&str`, so this needs to be handled as a special case (or this is is a general case, broader than Strings), and I need to find a way to compare `#property_type`.  Roughly:

```
if type is String:  // TBD how to cmp #property_type against String
  fn String -> &str
else:
  fn T -> T
```

[the remark this answers is not in the document]
Comment (alexjmacdonald) [reply src/rust/derive-dynamic-node/src/lib.rs:227]: I've been hacking at this last step, and I hit a wall.

I want to determine if `property_type` corresponds to a String.  However, we're operating on the AST, so while we know `String` is a Type node in the AST, we don't know that `String` and `std::String::String` and `type Gnirts = String` are all the same thing until we're trying to parse our finished dynamically-generated getter function.

What we _can_ do is look for `std::String::String` and `String`, and then say "oh okay, this getter returns &str."

Relevant StackOverflow on how to get a TypePath from a Type and operate on it:
https://stackoverflow.com/questions/55271857/how-can-i-get-the-t-from-an-optiont-when-using-syn

[the remark this answers is not in the document]
Comment (alexjmacdonald) [reply src/rust/derive-dynamic-node/src/lib.rs:227]:
```
    // Given the property type, determine:
    // - the method on `property` to call
    // - the the type of the above, which will be the return type of the function
    let (return_type, method_ident): (syn::Type, syn::Ident) = match property_type {
      Type::Path(typepath) => match typepath.path.segments.iter().into_iter().map(|x|x.ident.to_string()).collect::<Vec<String>>().join("::").as_ref() {
        // These are the type signatures we recognize (must be literal matches):
        "String"|"std::string::String" => (parse_quote!(&str), parse_quote!(as_str_prop)),
        "u64"                          => (parse_quote!(u64),  parse_quote!(as_uint_prop)),
        "i64"                          => (parse_quote!(i64),  parse_quote!(as_int_prop)),
        // Anything else no-ops out, without implementing a getter.
        _ => return implementation,
      },
      _ => panic!("Tried to dynamically construct getter for unrecognized type!"),
    };

    let get_method_implementation = quote!(
        fn #get_method_name(&self) -> Option<#return_type> {
            let property_result: Option<&NodeProperty> = self.get_dynamic_node().get_property(#property_name_str);

            match property_result {
              Some(ref property) => property. #method_ident(),
              None => None,
            }
        }
    );
```

This works?

[the remark this answers is not in the document]
Comment (wimax-grapl) [reply src/rust/derive-dynamic-node/src/lib.rs:227]: Yeah, this restriction makes sense I guess. No String type aliases are likely to be passed in here, I'd imagine.
Just comment the hell out of the panic i guess?

File: src/rust/graph-descriptions/src/dynamic_node.rs [hunk 3]
```diff
@@ -5,6 +5,10 @@ use crate::graph_description::{id_strategy, node_property, DynamicNode, IdStrate
 use crate::node::NodeT;
 
 impl DynamicNode {
+    pub fn get_property(&self, name: impl Into<String>) -> Option<&NodeProperty> {
```
Comment (colin-grapl) [inline src/rust/graph-descriptions/src/dynamic_node.rs:8]: I think this is better as
```
    pub fn get_property(&self, name: impl AsRef<str>) -> Option<&NodeProperty> {
        self.properties.get(name.as_ref())
    }
```

File: src/rust/graph-descriptions/src/lib.rs [hunk 4]
```diff
@@ -62,6 +62,27 @@ impl From<u64> for NodeProperty {
     }
 }
 
+impl From<NodeProperty> for Option<String> {
+    fn from(np: NodeProperty) -> Option<String> {
+        match np.as_str_prop() {
+            Some(str) => Some(String::from(str)),
```
Comment (colin-grapl) [inline src/rust/graph-descriptions/src/lib.rs:68]: Not worth changing, but I think shadowing 'str' is a bit strange, and I'd suggest just using 's' or something. Really minor nit though, nbd.

Also, in this case you can do:

```
np.as_str_prop().map(String::from)
```
and remove the match block entirely.

**Review by colin-grapl — APPROVED** [review]

**Pull request merged**